src/org/purefn/river/batch.clj

```diff
@@ -0,0 +1,137 @@
+(ns org.purefn.river.batch
+  "A consumer suitable for batch reading operations from Kafka."
+  (:require [clojure.pprint :as pprint]
+            [clojure.spec.alpha :as s]
+            [com.stuartsierra.component :as component]
+            ;; temp, remove
+            [org.purefn.river.serdes.nippy :as serdes]
+            ;;----
+            [taoensso.timbre :as log])
+  (:import [org.apache.kafka.clients.consumer KafkaConsumer]
+           [org.apache.kafka.common TopicPartition]
+           [org.apache.kafka.common.errors WakeupException]
+           [org.purefn.river Message]))
+
+(defn kafka-consumer
+  [group-id servers]
+  (KafkaConsumer. {"auto.offset.reset" "earliest"
+                   "bootstrap.servers" servers
+                   "enable.auto.commit" false
+                   "group.id" group-id
+                   "client.id" group-id}
+                  (serdes/nippy-deserializer)
+                  (serdes/nippy-deserializer)))
+
+(defn create-consumers
+  [{:keys [::topics ::group-id ::threads ::bootstrap-servers] :as config}]
+  (log/info config)
+  (->> (range threads)
+       (map (fn [_]
+              (log/info "Creating consumer" {:group-id group-id
+                                             :topics topics})
+                (doto (kafka-consumer group-id bootstrap-servers)
+                  (.subscribe topics))))))
+
+(defn process
+  [^KafkaConsumer consumer
+   ^clojure.lang.Atom closing
+   {:keys [::timeout] :as config}
+   dependencies
+   process-fn]
+  (let [commit #(.commitAsync consumer)
+        pfn (if (= 4 (->> (meta process-fn)
+                          (:arglists)
+                          (map count)
+                          (reduce max)))
+              (partial process-fn dependencies)
+              process-fn)]
+    (try
+      (loop [state {}]
+        (when-not @closing
+          (let [records (seq (.poll consumer timeout))
+                next-state (if records
+                             (pfn state
+                                  (map #(Message. %) records)
+                                  commit)
+                             state)]
+            (recur next-state))))
+
+      (catch WakeupException ex
+        (when-not @closing
+          (throw ex)))
+
+      (catch Exception ex
+        (log/error ex)
+        (throw ex))
+
+      (finally
+        (log/info "Closing consumer for"
+                  (->> (.assignment consumer)
+                       (map (juxt (memfn topic)
+                                  (memfn partition)))
+                       (pprint/pprint)
+                       (with-out-str)))
+        (.close consumer)))))
+
+(defrecord BatchConsumer
+    [config consumers process-fn]
+
+  component/Lifecycle
+  (start [this]
+    (assoc this
+           :consumers
+           (mapv (fn [consumer]
+                  (let [closing (atom false)]
```

Comment:
One alternative to this signal-atom strategy that I've used lately and liked is have `process` return a closing function. Could be something like:

```
(defn processor
 [consumer config etc etc]
 (let [f (future (loop [state] ...)]
    (fn []
      (log/info "closing etc etc")
      (future-cancel f))))
```